### PR TITLE
Native auth extensibility

### PR DESCRIPTION
**Pull request by Yuki-YuXin**
Include those commits
image

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/client/statemachine/results/SignUpResult.kt [hunk 1]
```diff
@@ -23,24 +23,21 @@
...
-import com.microsoft.identity.client.statemachine.UserAlreadyExistsError
 import com.microsoft.identity.client.statemachine.states.SignInAfterSignUpState
 import com.microsoft.identity.client.statemachine.states.SignUpAttributesRequiredState
 import com.microsoft.identity.client.statemachine.states.SignUpCodeRequiredState
 import com.microsoft.identity.client.statemachine.states.SignUpPasswordRequiredState
 
 /**
- * Sign up result, produced by
+ * Results for native sign up flows.
+ * TODO add documentation & links to learn.microsoft.com
```
Comment (SaurabhMSFT) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/results/SignUpResult.kt:33]: can you remove this todo?

File: msal/src/main/java/com/microsoft/identity/client/statemachine/states/SignUpStates.kt [hunk 2]
```diff
@@ -62,14 +63,6 @@ import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
 import java.io.Serializable
 
-/**
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/states/SignUpStates.kt:65]: Why were these comments removed?

File: msal/src/main/java/com/microsoft/identity/client/statemachine/states/SignUpStates.kt [hunk 3]
```diff
@@ -280,29 +269,18 @@ class SignUpCodeRequiredState internal constructor(
...
+                        exception = result.exception
                     )
                 }
             }
         }
     }
 }
 
-/**
- * Native Auth uses a state machine to denote state and transitions for a user.
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/states/SignUpStates.kt:299]: Same here

File: msal/src/main/java/com/microsoft/identity/client/statemachine/states/SignUpStates.kt [hunk 4]
```diff
@@ -454,14 +425,6 @@ class SignUpPasswordRequiredState internal constructor(
...
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/states/SignUpStates.kt:458]: And here

File: msal/src/main/java/com/microsoft/identity/client/statemachine/states/SignUpStates.kt [hunk 5]
```diff
@@ -563,42 +525,29 @@ class SignUpAttributesRequiredState internal constructor(
...
+                        exception = result.exception
                     )
                 }
             }
         }
     }
 }
 
-/**
- * Native Auth uses a state machine to denote state and transitions for a user.
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/states/SignUpStates.kt:595]: And herre

File: msal/src/main/java/com/microsoft/identity/client/INativeAuthPublicClientApplication.kt [hunk 6]
```diff
@@ -109,66 +109,15 @@ interface INativeAuthPublicClientApplication : IPublicClientApplication {
      */
     fun signInUsingPassword(username: String, password: CharArray, scopes: List<String>? = null, callback: NativeAuthPublicClientApplication.SignInUsingPasswordCallback)
 
-    /**
```
Comment (SaurabhMSFT) [inline msal/src/main/java/com/microsoft/identity/client/INativeAuthPublicClientApplication.kt:112]: Can you restore deleted comments from this file?

File: msal/src/main/java/com/microsoft/identity/client/NativeAuthPublicClientApplication.kt [hunk 7]
```diff
@@ -343,38 +323,53 @@ class NativeAuthPublicClientApplication(
                         )
                     )
                 }
-                is SignInCommandResult.InvalidCredentials -> {
+                is SignInCommandResult.Complete -> {
```
Comment (SaurabhMSFT) [inline msal/src/main/java/com/microsoft/identity/client/NativeAuthPublicClientApplication.kt:326]: We can merge the clauses for `Complete`, `InvalidCredentials` and `UnknownError` into `else` block. Or at least  `Complete`, `InvalidCredentials`  can be merged into a single else block.

File: msal/src/main/java/com/microsoft/identity/client/statemachine/results/ResetPasswordResult.kt [hunk 8]
```diff
@@ -23,18 +23,19 @@
...
-import com.microsoft.identity.client.statemachine.GeneralError
-import com.microsoft.identity.client.statemachine.IncorrectCodeError
-import com.microsoft.identity.client.statemachine.InvalidPasswordError
-import com.microsoft.identity.client.statemachine.UserNotFoundError
 import com.microsoft.identity.client.statemachine.states.ResetPasswordCodeRequiredState
 import com.microsoft.identity.client.statemachine.states.ResetPasswordPasswordRequiredState
 
+/**
+ * Results for native Self-service password reset flows.
+ * TODO add documentation & links to learn.microsoft.com
```
Comment (SaurabhMSFT) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/results/ResetPasswordResult.kt:31]: We should remove this TODO as the link to documentation has been added below.

File: msal/src/main/java/com/microsoft/identity/client/statemachine/states/BaseStates.kt [hunk 9]
```diff
@@ -23,9 +23,9 @@
...
 package com.microsoft.identity.client.statemachine.states
 
-sealed interface State
+interface State
 
 /**
  * BaseState is the base class for various states in the Native Auth state machine.
  */
-abstract class BaseState(internal open val flowToken: String?)
+abstract class BaseState(internal open val flowToken: String?)
```
Comment (SaurabhMSFT) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/states/BaseStates.kt:31]: New line should be added at the end.

File: msal/src/main/java/com/microsoft/identity/client/statemachine/states/ResetPasswordStates.kt [hunk 10]
```diff
@@ -55,13 +56,6 @@ import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
 import java.io.Serializable
 
-/**
```
Comment (SaurabhMSFT) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/states/ResetPasswordStates.kt:58]: Can you restore this comment

File: msal/src/main/java/com/microsoft/identity/client/statemachine/states/ResetPasswordStates.kt [hunk 11]
```diff
@@ -148,22 +142,17 @@ class ResetPasswordCodeRequiredState internal constructor(
...
+                        correlationId = result.correlationId,
+                        exception = result.exception
                     )
                 }
             }
         }
     }
 
-    /**
-     * ResendCodeCallback receives the result for submit code for Reset Password for Native Auth
```
Comment (SaurabhMSFT) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/states/ResetPasswordStates.kt:165]: Can you restore this comment

File: msal/src/main/java/com/microsoft/identity/client/statemachine/states/ResetPasswordStates.kt [hunk 12]
```diff
@@ -234,28 +224,18 @@ class ResetPasswordCodeRequiredState internal constructor(
...
+                        exception = result.exception
                     )
                 }
             }
         }
     }
 }
 
-/**
- * Native Auth uses a state machine to denote state and transitions for a user.
```
Comment (SaurabhMSFT) [inline msal/src/main/java/com/microsoft/identity/client/statemachine/states/ResetPasswordStates.kt:253]: Can you restore this comment?

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java [hunk 13]
```diff
@@ -258,7 +270,11 @@ public void onError(@NonNull BaseException exception) {
         };
         application.signInUsingPassword(username, password, null, callback);
         // 1a. Server returns invalid password error
-        assertTrue(signInResult.get(30, TimeUnit.SECONDS) instanceof SignInResult.InvalidCredentials);
+        SignInUsingPasswordResult result = signInResult.get(30, TimeUnit.SECONDS);
+        assertTrue(result instanceof SignInUsingPasswordError);
```
Comment (SaurabhMSFT) [inline msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java:274]: We should avoid using `assertTrue` where we are no comparing a boolean object. If `assertTrue` fails in this case then we won't know from the test run about the actual class  of `result`. Below shows an example using `assertThat`

https://stackoverflow.com/questions/12404650/assert-an-object-is-a-specific-type

**Review by SaurabhMSFT — CHANGES_REQUESTED** [review]

**Pull request closed without merging**